makelove/macos.py

```diff
@@ -0,0 +1,245 @@
+import io
+import os
+import shutil
+import struct
+import sys
+from datetime import datetime
+from zipfile import ZipFile
+from urllib.request import urlopen, urlretrieve, URLError
+
+from PIL import Image
+
+from .util import eprint, tmpfile
+from .windows import get_default_love_binary_dir, get_download_url
+
+
+def download_love(version, platform):
+    """
+    Note, mac builds are stored as zip files because extracting them
+    would lose data about symlinks when building on windows
+    """
+    target_path = get_default_love_binary_dir(version, platform)
+    print("Downloading love binaries to: '{}'".format(target_path))
+
+    os.makedirs(target_path, exist_ok=True)
+    try:
+        download_url = get_download_url(version, platform)
+        print("Downloading '{}'..".format(download_url))
+        with urlopen(download_url) as response:
+            with open(os.path.join(target_path, "love.zip"), 'wb') as target_f:
+                target_f.write(response.read())
+    except URLError as exc:
+        eprint("Could not download löve: {}".format(exc))
+        eprint(
+            "If there is in fact no download on GitHub for this version, specify 'love_binaries' manually."
+        )
+        sys.exit(1)
+    print("Download complete")
+
+
+def write_file(pkg, name, content):
+    if isinstance(pkg, str):
+        mode = "w" if isinstance(content, str) else "wb"
+        with open(name, mode) as f:
+            f.write(content)
+    elif isinstance(pkg, ZipFile):
+        pkg.writestr(name, content)
+
+
+def make_icns(iconfile, icon_image_file):
+    """
+    iconfile: an open file to write the ICNS file contents into (mode: wb)
+    icon_image: a PIL.Image object of the icon image
+
+    Based on code from learn-python.com:
+      https://learning-python.com/cgi/showcode.py?name=pymailgui-products/unzipped/build/build-icons/iconify.py
+    """
+    icon_image = Image.open(icon_image_file)
+
+    # must all be square (width=height) and of standard pixel sizes
+    width, height = icon_image.size  # a 2-tuple
+    if width != height:
+        raise ValueError('Invalid image size, discarded: %d x %d.' % (width, height))
```

Comment:
Also rather use `sys.exit` here

Reply:
I usually like exceptions because then you can call the function and handle the exception when you need similar functionality in the future. for this case sys.exit works fine though